dhis2/dhis2-core
# feat: add validation result deletion to API [DHIS2-7439]

### Summary
Adds `DELETE` HTTP methods to the validation result API.

* delete by ID: `DELETE /api/36/validationResults/<id>`
* delete by filters: `DELETE /api/36/validationResults?<filters>`

Filters parameters are:

* `ou=<uid>` : filter by organisation unit(s)
* `vr=<uid>`: filter by validation rule(s)
* `pe=<ISO-period>`: filter by the period the result is linked to
* `created=<ISO-period>`: filter by a period in which the result was created
* `notificationSent=<boolean>`: filter by status of the `notificationSent` flag of the results

Filters can use any combination but must use at least one filtering property. `ou` (organisation unit) and `vr` (validation rule) UIDs may occur more than once. `pe` and `created` (date) period can at most occur once and must refer to valid ISO date expressions. The notificationSent can be either `true` or `false` (or left out).

All matching validation results are deleted returning a `204` (no content) on success.
Both delete operations require the `F_PERFORM_MAINTENANCE` right (_Perform maintenance tasks_). If the user does not have this right a `403` (Forbidden) error is returned. 
If delete by ID is used on a non existing result a `404` (not found) is returned.

### Notes for Reviewers
Please not that the `@Transactional` annotation on `DefaultValidationResultService` was changed to `@Transactional( readOnly = true )` making all operations read only by default. Writing operations were annotated `@Transactional( readOnly = false )` explicitly. This affects the newly added delete operations as well as the `saveValidationResults`. This operation was manually tested when generating the validation results as described later.

### Automatic Testing
PR adds 

* new tests to test the delete database operations on store layer
* new tests to verify the validation performed for the request filter parameters in the service layer

### Manual Testing
On basis of Sierra Leone sample database:

1. clone fresh database
2. start application
3. goto _Data Quality_ app
4. select _Validation Rule Analysis_
5. change start date to 2020-01-01
6. check _Persist new results_ checkbox
7. select root organisation
8. click _Validate_

To test the API use a REST test tool, try for example the following URLs (with `DELETE` HTTP method)
Use GET for the same URLs to see the status before and after the delete.

* http://localhost:8080/api/36/validationResults/1497845 (query for all results first and pick an ID that exists)
* http://localhost:8080/api/36/validationResults?ou=jNb63DIHuwU

### Documentation
Can be found https://github.com/dhis2/dhis2-docs/pull/595
